File: lexicon/adapters/typescript/src/semantic-facts.ts
import * as ts from "typescript";
import { spanFor, staticTarget } from "./contract";
import type { FactStore, FileContext } from "./model";

const CAPABILITIES = "control-flow,error-handling,calls,source-spans,outcome-obligations";
type ErrorAction = "propagate" | "record" | "recover";

type ActionEvidence = {
  action: ErrorAction;
  node: ts.Node;
};

export function emitSemanticFacts(contexts: FileContext[], facts: FactStore): void {
  for (const context of contexts) {
    emitCapabilities(context, facts);
    visit(context.sourceFile, context, facts);
  }
}

function emitCapabilities(context: FileContext, facts: FactStore): void {
  const language = semanticLanguage(context);
  const identity = `@semantic/capabilities/${language}/${context.relativePath}`;
  facts.addNode(
    "protocol",
    `semantic-capabilities:${language}:${CAPABILITIES}`,
    context.relativePath,
    identity,
    identity,
  );
}

function visit(node: ts.Node, context: FileContext, facts: FactStore): void {
  if (ts.isCatchClause(node)) emitCatchClause(node, context, facts);
  ts.forEachChild(node, (child) => visit(child, context, facts));
}

function emitCatchClause(clause: ts.CatchClause, context: FileContext, facts: FactStore): void {
  const language = semanticLanguage(context);
  const start = clause.getStart(context.sourceFile);
  const identity = `@semantic/error-handler/${language}/${context.relativePath}:${start}`;
  const handlerId = facts.addNode(
    "protocol",
    `error-handler:${language}`,
    context.relativePath,
    identity,
    identity,
    spanFor(clause, context.sourceFile, context.relativePath),
  );
  for (const evidence of collectActions(clause.block)) {
    const actionStart = evidence.node.getStart(context.sourceFile);
    const actionIdentity = `${identity}/${evidence.action}:${actionStart}`;
    const actionId = facts.addNode(
      "protocol",
      `error-action:${evidence.action}`,
      context.relativePath,
      actionIdentity,
      actionIdentity,
      spanFor(evidence.node, context.sourceFile, context.relativePath),
    );
    facts.addEdge(handlerId, actionId, "contains", spanFor(evidence.node, context.sourceFile, context.relativePath));
  }
}

function collectActions(block: ts.Block): ActionEvidence[] {
  const actions = new Map<ErrorAction, ts.Node>();
  const record = (action: ErrorAction, node: ts.Node): void => {
    if (!actions.has(action)) actions.set(action, node);
  };
  const inspect = (node: ts.Node): void => {
    if (node !== block && (ts.isFunctionLike(node) || ts.isCatchClause(node))) return;
    if (ts.isThrowStatement(node)) record("propagate", node);
    if (ts.isReturnStatement(node) && node.expression) {
      if (isPromiseReject(node.expression)) record("propagate", node);
      else record("recover", node);
    }
    if (ts.isCallExpression(node)) {
      const target = staticTarget(node.expression);
      if (target === "Promise.reject") record("propagate", node);
      else if (isRecordingTarget(target)) record("record", node);
      else record("recover", node);
    }
    if (ts.isBinaryExpression(node) && isAssignmentOperator(node.operatorToken.kind)) record("recover", node);
    if (ts.isBreakStatement(node) || ts.isContinueStatement(node)) record("recover", node);
    ts.forEachChild(node, inspect);
  };
  inspect(block);
  return [...actions.entries()].map(([action, node]) => ({ action, node }));
}

function isPromiseReject(expression: ts.Expression): boolean {
  return ts.isCallExpression(expression) && staticTarget(expression.expression) === "Promise.reject";
}

function isRecordingTarget(target: string | null): boolean {
  if (!target) return false;
  const leaf = target.split(".").at(-1)?.toLowerCase() ?? "";
  return ["error", "warn", "log", "reporterror", "captureerror", "captureexception"].includes(leaf);
}

function isAssignmentOperator(kind: ts.SyntaxKind): boolean {
  return kind >= ts.SyntaxKind.FirstAssignment && kind <= ts.SyntaxKind.LastAssignment;
}

export function semanticLanguage(context: FileContext): "javascript" | "typescript" {
  return /\.(?:js|jsx|mjs|cjs)$/i.test(context.relativePath) ? "javascript" : "typescript";
}

File: lexicon/adapters/rust/src/semantic_facts.rs
use crate::model::{Context, SourceFile};
use crate::paths::{span_start, span_value};
use proc_macro2::Span;
use std::collections::BTreeMap;
use syn::spanned::Spanned;
use syn::visit::{self, Visit};

const CAPABILITIES: &str = "control-flow,error-handling,calls,source-spans,outcome-obligations";

use crate::semantic_actions::{actions_for_block, actions_for_expr, ErrorAction};

struct Handler {
    span: Span,
    actions: Vec<(ErrorAction, Span)>,
}

pub(crate) fn emit(context: &mut Context) {
    let sources: Vec<SourceFile> = context.sources.values().cloned().collect();
    for source in sources {
        emit_capabilities(context, &source);
        let mut collector = HandlerCollector::default();
        collector.visit_file(&source.syntax);
        for handler in collector.handlers {
            emit_handler(context, &source, handler);
        }
    }
}

fn emit_capabilities(context: &mut Context, source: &SourceFile) {
    let identity = format!("@semantic/capabilities/rust/{}", source.relative);
    context.facts.add_node(
        "rust",
        "protocol",
        &identity,
        &format!("semantic-capabilities:rust:{CAPABILITIES}"),
        &source.relative,
        &identity,
        None,
        None,
        BTreeMap::new(),
    );
}

fn emit_handler(context: &mut Context, source: &SourceFile, handler: Handler) {
    let (line, column) = span_start(handler.span);
    let identity = format!(
        "@semantic/error-handler/rust/{}:{line}:{column}",
        source.relative
    );
    let handler_id = context.facts.add_node(
        "rust",
        "protocol",
        &identity,
        "error-handler:rust",
        &source.relative,
        &identity,
        None,
        span_value(handler.span, &source.relative),
        BTreeMap::new(),
    );
    for (action, span) in handler.actions {
        let (action_line, action_column) = span_start(span);
        let action_identity = format!("{identity}/{}:{action_line}:{action_column}", action.name());
        let action_id = context.facts.add_node(
            "rust",
            "protocol",
            &action_identity,
            &format!("error-action:{}", action.name()),
            &source.relative,
            &action_identity,
            None,
            span_value(span, &source.relative),
            BTreeMap::new(),
        );
        context.facts.add_edge(
            &handler_id,
            &action_id,
            "contains",
            span_value(span, &source.relative),
        );
    }
}

#[derive(Default)]
struct HandlerCollector {
    handlers: Vec<Handler>,
}

impl<'ast> Visit<'ast> for HandlerCollector {
    fn visit_expr_match(&mut self, node: &'ast syn::ExprMatch) {
        for arm in &node.arms {
            if is_err_pattern(&arm.pat) {
                self.handlers.push(Handler {
                    span: arm.span(),
                    actions: actions_for_expr(&arm.body),
                });
            }
        }
        visit::visit_expr_match(self, node);
    }

    fn visit_expr_if(&mut self, node: &'ast syn::ExprIf) {
        if let syn::Expr::Let(condition) = node.cond.as_ref() {
            if is_err_pattern(&condition.pat) {
                self.handlers.push(Handler {
                    span: node.then_branch.span(),
                    actions: actions_for_block(&node.then_branch),
                });
            }
        }
        visit::visit_expr_if(self, node);
    }
}

fn is_err_pattern(pattern: &syn::Pat) -> bool {
    match pattern {
        syn::Pat::TupleStruct(value) => value
            .path
            .segments
            .last()
            .is_some_and(|part| part.ident == "Err"),
        syn::Pat::Path(value) => value
            .path
            .segments
            .last()
            .is_some_and(|part| part.ident == "Err"),
        syn::Pat::Or(value) => value.cases.iter().any(is_err_pattern),
        syn::Pat::Paren(value) => is_err_pattern(&value.pat),
        syn::Pat::Reference(value) => is_err_pattern(&value.pat),
        syn::Pat::Type(value) => is_err_pattern(&value.pat),
        _ => false,
    }
}
